types/src/to_abi.rs

```diff
@@ -0,0 +1,446 @@
+    Option(Option<Name>, Box<Abi>),
+    Account20(Option<Name>),
+    Account32(Option<Name>),
+    H256(Option<Name>),
+    Bytes(Option<Name>),
+    Value256(Option<Name>),
+    Value128(Option<Name>),
+    Value64(Option<Name>),
+    Value32(Option<Name>),
+    Byte(Option<Name>),
+    Bool(Option<Name>),
+    Vec(Option<Name>, Box<Abi>),
+    Tuple(Option<Name>, (Box<Abi>, Box<Abi>)),
+}
+
+impl Abi {
+    pub fn get_name(&self) -> Option<Name> {
+        match self {
+            Abi::Struct(name, _) => name.clone(),
+            Abi::Log(name, _) => name.clone(),
+            Abi::Enum(name, _) => name.clone(),
+            Abi::Option(name, _) => name.clone(),
+            Abi::Account20(name) => name.clone(),
+            Abi::Account32(name) => name.clone(),
+            Abi::H256(name) => name.clone(),
+            Abi::Bytes(name) => name.clone(),
+            Abi::Value256(name) => name.clone(),
+            Abi::Value128(name) => name.clone(),
+            Abi::Value64(name) => name.clone(),
+            Abi::Value32(name) => name.clone(),
+            Abi::Byte(name) => name.clone(),
+            Abi::Bool(name) => name.clone(),
+            Abi::Vec(name, _) => name.clone(),
+            Abi::Tuple(name, _) => name.clone(),
+        }
+    }
+
+    pub fn get_type_size(&self) -> usize {
+        match self {
+            Abi::Struct(_, _fields) => 1,
+            Abi::Log(_, _fields) => 1,
+            Abi::Enum(_, _fields) => 1,
+            Abi::Option(_, _field) => 1,
+            Abi::Account20(_) => 20,
+            Abi::Account32(_) => 32,
+            Abi::H256(_name) => 32,
+            Abi::Bytes(_) => 32,
+            Abi::Value256(_) => 32,
+            Abi::Value128(_) => 16,
+            Abi::Value64(_) => 8,
+            Abi::Value32(_) => 4,
+            Abi::Byte(_) => 1,
+            Abi::Bool(_) => 1,
+            Abi::Vec(_, _field) => 1,
+            Abi::Tuple(_, (_field1, _field2)) => 1,
+        }
+    }
+
+    pub fn get_size(&self) -> usize {
+        match self {
+            Abi::Struct(_, fields) => fields.iter().map(|f| f.get_size()).sum::<usize>(),
+            Abi::Log(_, fields) => fields.iter().map(|f| f.get_size()).sum::<usize>(),
+            Abi::Enum(_, fields) => fields.iter().map(|f| f.get_size()).sum::<usize>(),
+            Abi::Option(_, field) => 1 + field.get_size(),
+            Abi::Account20(_) => 20,
+            Abi::Account32(_) => 32,
+            Abi::H256(_) => 32,
+            Abi::Bytes(_) => 32,
+            Abi::Value256(_) => 32,
+            Abi::Value128(_) => 16,
+            Abi::Value64(_) => 8,
+            Abi::Value32(_) => 4,
+            Abi::Byte(_) => 1,
+            Abi::Bool(_) => 1,
+            // this needs to be multiplied by the length of the vec
+            Abi::Vec(_, field) => 1usize + field.get_size(),
+            Abi::Tuple(_, (field1, field2)) => field1.get_size() + field2.get_size(),
+        }
+    }
+
+    pub fn decode_topics_as_rlp(
```

Comment:
There is a lot of unwrapping here, and copying the data.

I would suggest taking the `input` as a `bytes::BytesMut` and stripping the slice as we take bits from it. Here we are cloning a fair amount.

Reply:
See here for an example: https://github.com/t3rn/t3rn/blob/a665590c85e5d0c7b343bc52acd848469d6a50df/types/src/side_effect.rs#L217

Reply:
it's done, answered to following comment.